util/sli/sli.go

```diff
@@ -0,0 +1,119 @@
+// Copyright 2021 PingCAP, Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sli
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/tidb/metrics"
+)
+
+// TxnWriteThroughputSLI uses to report transaction write throughput metrics for SLI.
+type TxnWriteThroughputSLI struct {
+	invalid   bool
+	affectRow uint64
+	writeSize int
+	readKeys  int
+	writeKeys int
+	writeTime time.Duration
+}
+
+// FinishExecuteStmt records the cost for write statement which affect rows more than 0.
+// And report metrics when the transaction is committed.
+func (t *TxnWriteThroughputSLI) FinishExecuteStmt(cost time.Duration, affectRow uint64, inTxn bool) {
+	if affectRow > 0 {
+		t.writeTime += cost
+		t.affectRow += affectRow
+	}
+
+	// Currently not in transaction means the last transaction is finish, should report metrics and reset data.
+	if !inTxn {
+		if affectRow == 0 {
+			// AffectRows is 0 when statement is commit.
+			t.writeTime += cost
+		}
+		// Report metrics after commit this transaction.
+		t.reportMetric()
+
+		// Skip reset for test.
+		failpoint.Inject("CheckTxnWriteThroughput", func() {
+			failpoint.Return()
+		})
+
+		// Reset for next transaction.
+		t.Reset()
+	}
+}
+
+// AddReadKeys adds the read keys.
+func (t *TxnWriteThroughputSLI) AddReadKeys(readKeys int64) {
+	t.readKeys += int(readKeys)
+}
+
+// CommittedTxn marks the transaction is committed and record the write size and keys.
+func (t *TxnWriteThroughputSLI) CommittedTxn(size, keys int) {
```

Comment:
Bad function name.

Reply:
How about rename to `CommitTxn`?

Reply:
Is it related to `commit`?  If I want to search something about `commit a transaction`, I would found this function.

Reply:
nice catch, maybe rename to `AddTxnSize` is better.

Reply:
FinishExecuteTxn?